scanner.c

```diff
@@ -143,68 +231,304 @@ scanner_process(const char * line, struct scanner_statistics * data) {
-	}
-	else if(sc_stat == -1 && cc_stat == 0) {
+	} else if (sc_stat == -1 && cc_stat == 0) {
 		data->scan_duration_cc_0_count++;
 		data->scan_duration_cc_0_sum += duration;
-	}
-	else if(sc_stat == -1 && cc_stat == 1) {
+	} else if (sc_stat == -1 && cc_stat == 1) {
 		data->scan_duration_cc_1_count++;
 		data->scan_duration_cc_1_sum += duration;
-	}
-	else if(sc_stat == 0 && cc_stat == -1) {
+	} else if (sc_stat == 0 && cc_stat == -1) {
 		data->scan_duration_sc_0_count++;
 		data->scan_duration_sc_0_sum += duration;
-	}
-	else if(sc_stat == 0 && cc_stat == 0) {
+	} else if (sc_stat == 0 && cc_stat == 0) {
 		data->scan_duration_sc_0_cc_0_count++;
 		data->scan_duration_sc_0_cc_0_sum += duration;
-	}
-	else if(sc_stat == 0 && cc_stat == 1) {
+	} else if (sc_stat == 0 && cc_stat == 1) {
 		data->scan_duration_sc_0_cc_1_count++;
 		data->scan_duration_sc_0_cc_1_sum += duration;
-	}
-	else if(sc_stat == 1 && cc_stat == -1) {
+	} else if (sc_stat == 1 && cc_stat == -1) {
 		data->scan_duration_sc_1_count++;
 		data->scan_duration_sc_1_sum += duration;
-	}
-	else if(sc_stat == 1 && cc_stat == 0) {
+	} else if (sc_stat == 1 && cc_stat == 0) {
 		data->scan_duration_sc_1_cc_0_count++;
 		data->scan_duration_sc_1_cc_0_sum += duration;
-	}
-	else if(sc_stat == 1 && cc_stat == 1) {
+	} else if (sc_stat == 1 && cc_stat == 1) {
 		data->scan_duration_sc_1_cc_1_count++;
 		data->scan_duration_sc_1_cc_1_sum += duration;
 	}
+
+	/* Just one detected error is enough for an evidence and eventual alert */
+	int cur_field_num = 4;
+	for (; cur_field_num <= NUM_OF_FIELDS; cur_field_num++) {
+		line = get_next_field(buf, sizeof buf, line + 1, '\t');
+		if (cur_field_num < NUM_OF_FIELDS && line == NULL) {
+			data->incorrectNumOfClmns = 1;
+			return;
+		} else if (cur_field_num >= 11) {
+			if (!*buf) {
+				data->empty_field = 1;
+				return;
+			} else if (strstr(buf, "NULL")) {
+				data->null_field = 1;
+				return;
+			}
+		}
+	}
+	if (line != NULL) {
+		data->incorrectNumOfClmns = 1;
+	}
 }
 
+#define STARTSWITH(str, start) (strncmp(start, str, sizeof start - 1) == 0)
+
 static
 int
-scanner_print_hdr(const char * name) {
-	nd_chart("scannerd", name, "type", "", "", "volume", "scannerd", "scannerd.scannerd_type", ND_CHART_TYPE_STACKED);
+get_conn_ip(const char * line, char * ip_lastpart) {
+	#define UNTOCONN "unable to connect to "
+	char ip[64];
+
+	ip_lastpart[0] = '\0';
+
+	if (!STARTSWITH(line, UNTOCONN)) return 0;
+	if ((line = get_next_field(ip, sizeof ip, line + sizeof UNTOCONN - 1, ' ')) == NULL) {
+		return 0;
+	}
+
+	// Get rid of the last '"' and ':'
+	int ipend = strlen(ip) - 3;
+
+	if (ipend < 4) return 1;
+
+	int ipstart = ipend;
+	for (; ipstart > ipend - 4 && ip[ipstart] != '.' && ip[ipstart] != ':'; ipstart--);
+
+	int ip_lastpart_len = ipend - ipstart;
+	ipstart++;
+	memcpy(ip_lastpart, ip + ipstart, ip_lastpart_len);
+	ip_lastpart[ip_lastpart_len] = '\0';
+	return 1;
+}
+
+static
+int
+get_scan_ip(const char * line, char * ip_lastpart) {
```

Comment:
This function duplicates `get_conn_ip`, it would be better to have just one function. I don't understand what benefit should `STARTSWITH`, `SCANWITH` and `UNTOCON` macros bring.